### ISSUE TITLE
Add WS support to cryo

### ISSUE THREAD

**Issue by libevm** [paradigmxyz/cryo#65]
add ws support pls 🙏

```
--ws OR --rpc
```

**Comment by 0xprames** [comment]
hey there - C/P'ing my messages from the TG on this

0xprames, [Sep 24, 2023 at 10:22:40 AM]:
yeah im going through the code + your PRs adding RetryClient (both the initial and the final one using ethers)

Looking at the alchemy docs for it rn and I see this w.r.t Websocket transport errors and how clients would do retries (as opposed to Http clients catching a 429 and retrying)

I think the first option you mention makes sense to unblock anyone who NEEDS a WS client asap (just creating a basic WS ethers provider w/o retry)

which confirms what Libevm was saying earlier re: no Ws support rn in ethers::RetryClient

 I suspect a WsRateLimitRetryPolicy may be a bit more complex to implement given that alchemy wss screencap above.

**Comment by 0xprames** [comment]
tl;dr  (deleted those messages bc I spammed chat)

-  i was asking why users may want WS support given Fetcher is only making jsonrpc request rn (discussed this further in the TG and came to the conclusion that it would be faster)
- I was saying as per alchemy docs, WS retry may be a bit more complicated than Http retries
- confirmed what you mentioned in the TG that ethers::RetryClient does not currently support Ws:

```
    let ws_provider = Provider::new(RetryClient::new(
        Ws::connect(Url::parse(ws_url).unwrap()).await.map_err(|_e|ParseError::ParseError("could not connect to ws_provider".to_string()))?,
        Box::new(HttpRateLimitRetryPolicy),
        args.max_retries,
        args.initial_backoff,
    ))
```

gives:
```
error[E0277]: the trait bound `ethers::ethers_providers::HttpRateLimitRetryPolicy: RetryPolicy<WsClientError>` is not satisfied
  --> crates/cli/src/parse/source.rs:22:9
   |
22 |         Box::new(HttpRateLimitRetryPolicy),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `RetryPolicy<WsClientError>` is not implemented for `ethers::ethers_providers::HttpRateLimitRetryPolicy`
   |
   = help: the trait `RetryPolicy<ethers::ethers_providers::HttpClientError>` is implemented for `ethers::ethers_providers::HttpRateLimitRetryPolicy`
   = note: required for the cast from `ethers::ethers_providers::HttpRateLimitRetryPolicy` to the object type `dyn RetryPolicy<WsClientError>`
```

**Comment by 0xprames** [comment]
so i think the solution @ipatka mentioned in that TG makes sense, which is to add a basic WS client without retries, and maybe iterate on that solution (either here or in ethers.rs codebase), adding retries eventually
